azurerm/resource_arm_image.go

```diff
@@ -0,0 +1,417 @@
+						},
+
+						"size_gb": {
+							Type:     schema.TypeInt,
+							Optional: true,
+						},
+					},
+				},
+			},
+
+			"tags": tagsSchema(),
+		},
+	}
+}
+
+func resourceArmImageCreateUpdate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*ArmClient)
+	imageClient := client.imageClient
+
+	log.Printf("[INFO] preparing arguments for AzureRM Image creation.")
+
+	name := d.Get("name").(string)
+	location := d.Get("location").(string)
+	resGroup := d.Get("resource_group_name").(string)
+	tags := d.Get("tags").(map[string]interface{})
+	expandedTags := expandTags(tags)
+	properties := compute.ImageProperties{}
+
+	osDisk, err := expandAzureRmImageOsDisk(d)
+	if err != nil {
+		return err
+	}
+
+	dataDisks, err := expandAzureRmImageDataDisks(d)
+	if err != nil {
+		return err
+	}
+
+	storageProfile := compute.ImageStorageProfile{
+		OsDisk:    osDisk,
+		DataDisks: &dataDisks,
+	}
+
+	sourceVM := compute.SubResource{}
+	if v, ok := d.GetOk("source_virtual_machine_id"); ok {
+		vmID := v.(string)
+		sourceVM = compute.SubResource{
+			ID: &vmID,
+		}
+	}
+
+	//either source VM or storage profile can be specified, but not both
+	if (compute.SubResource{}) == sourceVM {
+		//if both sourceVM and storageProfile are empty, return an error
+		if (compute.ImageStorageProfile{}) == storageProfile {
+			return fmt.Errorf("[ERROR] Cannot create image when both source VM and storage profile are empty")
+		}
+
+		properties = compute.ImageProperties{
+			StorageProfile: &storageProfile,
+		}
+	} else {
+		//creating an image from source VM
+		properties = compute.ImageProperties{
+			SourceVirtualMachine: &sourceVM,
+		}
+	}
+
+	createImage := compute.Image{
+		Name:            &name,
+		Location:        &location,
+		Tags:            expandedTags,
+		ImageProperties: &properties,
+	}
+
+	_, imageErr := imageClient.CreateOrUpdate(resGroup, name, createImage, make(chan struct{}))
+	err = <-imageErr
+	if err != nil {
+		return err
+	}
+
+	read, err := imageClient.Get(resGroup, name, "")
+	if err != nil {
+		return err
+	}
+	if read.ID == nil {
+		return fmt.Errorf("[ERROR] Cannot read AzureRM Image %s (resource group %s) ID", name, resGroup)
+	}
+
+	d.SetId(*read.ID)
+
+	return resourceArmImageRead(d, meta)
+}
+
+func resourceArmImageRead(d *schema.ResourceData, meta interface{}) error {
+	imageClient := meta.(*ArmClient).imageClient
+
+	id, err := parseAzureResourceID(d.Id())
+	if err != nil {
+		return err
+	}
+	resGroup := id.ResourceGroup
+	name := id.Path["images"]
+
+	resp, err := imageClient.Get(resGroup, name, "")
+	if err != nil {
+		if resp.StatusCode == http.StatusNotFound {
+			d.SetId("")
+			return nil
+		}
+		return fmt.Errorf("[ERROR] Error making Read request on AzureRM Image %s (resource group %s): %+v", name, resGroup, err)
+	}
+
+	d.Set("name", resp.Name)
+	d.Set("resource_group_name", resGroup)
+	d.Set("location", resp.Location)
+
+	if resp.SourceVirtualMachine != nil {
+		if resp.SourceVirtualMachine.ID != nil {
```

Comment:
I think we can remove this particular check `*.ID` as it should be handled by d.Set?